scene/gui/graph_edit.cpp

```diff
@@ -1646,6 +1646,585 @@ HBoxContainer *GraphEdit::get_zoom_hbox() {
+				if (!l.has(current_layer)) {
+					l.set(current_layer, Vector<StringName>{});
+				}
+				selected = true;
+				t.append_array(l[current_layer]);
+				l.set(current_layer, t);
+				Set<StringName> V;
+				V.insert(E->get());
+				set_operations(GraphEdit::UNION, u, V);
+			}
+		}
+		if (!selected) {
+			current_layer++;
+			set_operations(GraphEdit::UNION, z, u);
+		}
+		selected = false;
+	}
+
+	return l;
+}
+
+Vector<StringName> GraphEdit::split(const Vector<StringName> &l, const HashMap<StringName, Dictionary> &c) {
+	if (!l.size()) {
+		return Vector<StringName>();
+	}
+
+	StringName p = l[Math::random(0, l.size() - 1)];
+	Vector<StringName> left;
+	Vector<StringName> right;
+
+	for (int i = 0; i < l.size(); i++) {
+		if (p != l[i]) {
+			StringName q = l[i];
+			int cross_pq = c[p][q];
+			int cross_qp = c[q][p];
+			if (cross_pq > cross_qp) {
+				left.push_back(q);
+			} else {
+				right.push_back(q);
+			}
+		}
+	}
+
+	left.push_back(p);
+	left.append_array(right);
+	return left;
+}
+
+void GraphEdit::horizontal_alignment(Dictionary &root, Dictionary &align, const HashMap<int, Vector<StringName>> &l, const HashMap<StringName, Set<StringName>> &u, const Set<StringName> &s) {
+	//Initialising root & align
+	for (const Set<StringName>::Element *E = s.front(); E; E = E->next()) {
+		root[E->get()] = E->get();
+		align[E->get()] = E->get();
+	}
+
+	if (l.size() == 1) {
+		return;
+	}
+
+	for (unsigned int i = 1; i < l.size(); i++) {
+		Vector<StringName> lower_layer = l[i];
+		Vector<StringName> upper_layer = l[i - 1];
+		int r = -1;
+
+		for (int j = 0; j < lower_layer.size(); j++) {
+			Vector<Pair<int, StringName>> up;
+			StringName current_node = lower_layer[j];
+			for (int k = 0; k < upper_layer.size(); k++) {
+				StringName adjacent_neighbour = upper_layer[k];
+				if (u[current_node].has(adjacent_neighbour)) {
+					up.push_back(Pair<int, StringName>(k, adjacent_neighbour));
+				}
+			}
+
+			int start = up.size() / 2;
+			int end = up.size() % 2 ? start : start + 1;
+			for (int p = start; p <= end; p++) {
+				StringName Align = align[current_node];
+				if (Align == current_node && r < up[p].first) {
+					align[up[p].second] = lower_layer[j];
+					root[current_node] = root[up[p].second];
+					align[current_node] = root[up[p].second];
+					r = up[p].first;
+				}
+			}
+		}
+	}
+
+	//Print layers here
+	//Check if vectors of hashmap are being written
+	/*for (int i = 0; l.has(i); i++) {
+		for (int j = 0; j < l[i].size(); j++) {
+			StringName temp = l[i][j];
+		}
+	}
+	//Test for block formation
+	for (const Set<StringName>::Element *E = s.front(); E; E = E->next()) {
+		const StringName Root = root[E->get()];
+		const StringName Align = align[E->get()];
+	}*/
+}
+
+void GraphEdit::crossing_minimisation(HashMap<int, Vector<StringName>> &l, const HashMap<StringName, Set<StringName>> &u) {
+	if (l.size() == 1) {
+		return;
+	}
+
+	//Print layers here
+	//Check if vectors of hashmap are being written
```

Comment:
When the pr is ready to be submitted, make tests or delete the commented code because they cause confusion.

Reply:
Sure!